File: server/csharp/MetaObjects.Render/Recover/RecoverSchema.cs
namespace MetaObjects.Render.Recover;

/// <summary>
/// Top-level recover descriptor.
/// <see cref="RootName"/> is the XML root element tag / logical JSON root name.
/// <see cref="Fields"/> is never null; a null argument is normalised to an empty list.
/// </summary>
public sealed record RecoverSchema
{
    public Format Format { get; init; }
    public string RootName { get; init; }
    public IReadOnlyList<FieldSpec> Fields { get; init; }

    public RecoverSchema(Format format, string rootName, IReadOnlyList<FieldSpec>? fields = null)
    {
        Format = format;
        RootName = rootName;
        Fields = fields ?? Array.Empty<FieldSpec>();
    }
}

File: server/java/render/src/main/java/com/metaobjects/render/recover/RecoverSchema.java
package com.metaobjects.render.recover;

import java.util.List;
import java.util.Objects;

/** Top-level recover descriptor. rootName = the XML root tag / logical JSON root name. */
public record RecoverSchema(Format format, String rootName, List<FieldSpec> fields) {
    public RecoverSchema {
        Objects.requireNonNull(format, "format");
        Objects.requireNonNull(rootName, "rootName");
        fields = fields == null ? List.of() : List.copyOf(fields);
    }
}
